modules/sr/robot3/arduino.py

```diff
@@ -0,0 +1,289 @@
+    serial_number = '1234567890'
+    return MappingProxyType({
+        serial_number: Arduino(webot, serial_number),
+    })
+
+
+class Arduino:
+    """
+    The Arduino board interface.
+
+    This is intended to be used with Arduino Uno boards running the sbot firmware.
+
+    :param webot: The robot object to connect to devices on.
+    :param serial_number: The serial number of the board.
+    """
+    __slots__ = ('_serial_num', '_serial', '_pins', '_identity')
+
+    @staticmethod
+    def get_board_type() -> str:
+        """
+        Return the type of the board.
+
+        :return: The literal string 'Arduino'.
+        """
+        return 'Arduino'
+
+    def __init__(
+        self,
+        webot: Robot,
+        serial_number: str,
+    ) -> None:
+        # Stored for use in the identify method and repr.
+        self._serial_num = serial_number
+
+        self._pins = (
+            # Pins 0 and 1 are reserved for serial comms
+            Pin(0, supports_analog=False, disabled=True),
+            Pin(1, supports_analog=False, disabled=True),
+            # splat in both the digital and analog pins
+            *(Pin(index, supports_analog=False) for index in range(2, 14)),
+            *(Pin(index, supports_analog=True) for index in AnalogPins),
+        )
+
+        # Connect the devices to the pins.
+        # NOTE: the names here correspond to the names given to devices in Webots
+        # and, in some places, the keyboard controller.
+        self._pins[2]._connect_device(Microswitch(webot, 'back bump sensor'))
+        self._pins[3]._connect_device(Led(webot, 'led 1', pin_num=3))
+        self._pins[4]._connect_device(Led(webot, 'led 2', pin_num=4))
+        self._pins[AnalogPins.A0]._connect_device(DistanceSensor(webot, 'Front Left DS'))
+        self._pins[AnalogPins.A1]._connect_device(DistanceSensor(webot, 'Front Right DS'))
+        self._pins[AnalogPins.A2]._connect_device(DistanceSensor(webot, 'Left DS'))
+        self._pins[AnalogPins.A3]._connect_device(DistanceSensor(webot, 'Right DS'))
+        self._pins[AnalogPins.A4]._connect_device(DistanceSensor(webot, 'Front DS'))
+        self._pins[AnalogPins.A5]._connect_device(DistanceSensor(webot, 'Back DS'))
+        # NOTE: these pins don't exist on a real arduino.
+        self._pins[AnalogPins.A6]._connect_device(
+            PressureSensor(webot, 'finger pressure left'),
+        )
+        self._pins[AnalogPins.A7]._connect_device(
+            PressureSensor(webot, 'finger pressure right'),
+        )
+
+    def identify(self) -> BoardIdentity:
+        """
+        Get the identity of the board.
+
+        The asset tag of the board is the serial number.
+
+        :return: The identity of the board.
+        """
+
+        return BoardIdentity(
+            manufacturer="Student Robotics",
+            board_type="SRduino",
+            asset_tag=self._serial_num,
+            sw_version='1.0',
+        )
```

Comment:
This is redundant in the simulator. (It's also kinda wrong)
```suggestion
```

Reply:
This function is present in the kit API and returns a suitable value here to be compatible. This a step towards students being able to use almost the same code in the simulator as on the physical robot without it crashing due to undefined functions.

Reply:
I appreciate the desire to enable similar code between the two environments, however we should also bear in mind that any code which we have comes with a cost. This is especially true for code which is untested (as the majority of the code in the simulator unfortunately is). I would greatly prefer that we not introduce code which is not actually necessary to the operation of the simulator nor documented as part of the kit API.

Aiming for the entire code surface between the two environments being the same seems to me like it's going to create far more opportunities for things to get out of step than it will aid.

Unless the plan is also to update the kit API to include all the utility functions which the simulator needs/has/provides?

Reply:
One of the goals of this project was to bring into line the physical kit and simulator interfaces. To do this we should be including all the public interfaces accessable to the user. 
These functions are very basic and stub out the functions. They may be unnessasery but it will stop the students code from crashing if they run code between the robot and the simulator, I believe this to be valuable addition for a UX perspective. 

I think this approach of taking the kit API code and swapping the serial device out for a webots device is a good method. If the structure remains similar between the two, it will ease the development process.

Reply:
I also think this:

> It's also kinda wrong

Comes off as condesending, please in the future could it be replaced with a clear explaination of what you mean. I for one do not understand what you are refering to.

Reply:
> One of the goals of this project was to bring into line the physical kit and simulator interfaces.

I disagree with the depth that this is being taken to. Perhaps I'm mistaken, but it doesn't match understanding of the expectations we currently have for the simulator. I'd be happy to chat about this, however I would rather that this not become a distraction from getting the simulator to a point where it's usable for SR2024.

> > It's also kinda wrong
> 
> Also comes off as condesending, please in the future could it be replaced with a clear explaination of what you mean. I for one do not understand what you are refering to.

Fair.

> a clear explaination of what you mean.

The properties presented in the snippet I've suggested be removed don't make sense in the simulator. There is no board, so there is no manufacturer, no software to have a version of nor asset tag which goes with it. If we are going to present mocks of these values in the simulator (which is perhaps reasonable if we really want the API here to be exhaustive, though I don't think that's a good idea) then we should prefer to make them actually distinct from the real kits -- this is the approach taken by the existing metadata APIs where they exist (though mostly they don't).